File: LectureCodes/NumQuad/adaptquad/Matlab/adaptquad.m
function I = adaptquad(f,M,rtol,abstol)
% adaptive numerical quadrature: \texttt{f} is a function handle to integrand
h = diff(M);                     % distances of quadrature nodes \label{aq:1}
mp = 0.5*(M(1:end-1)+M(2:end));  % positions of midpoints \label{aq:2}
fx = f(M); fm = f(mp);           % \label{aq:3}
trp_loc  = h.*(fx(1:end-1)+2*fm+fx(2:end))/4; % trapezoidal rule \eqref{eq:comptrap}\label{aq:4}
simp_loc = h.*(fx(1:end-1)+4*fm+fx(2:end))/6; % Simpson rule \eqref{eq:compsimp}\label{aq:5}
I = sum(simp_loc);               % Simpson approximation of integral value\label{aq:6}
est_loc  = abs(simp_loc -trp_loc);  % local error estimate \eqref{eq:aqest}\label{aq:7}
err_tot  = sum(est_loc);            % estimate for quadrature error\label{aq:8}
% Termination based on \eqref{eq:aqstop}
if ((err_tot > rtol*abs(I)) && (err_tot > abstol)) % \label{aq:term}
  refcells = find(est_loc > 0.9*sum(est_loc)/length(h));
  I = adaptquad(f,sort([M,mp(refcells)]),rtol,abstol); % \label{aq:10} 
end

File: LectureCodes/NumQuad/adaptquad/Python/adaptquad.py
import numpy as np


def adaptquad(f, m, rtol, atol):
    """ Adaptive numerical quadrature """
    h = np.diff(m)  # Distances of quadrature nodes
    mp = 0.5 * (m[:-1] + m[1:])  # Positions of midpoints
    fx = f(m)
    fm = f(mp)
    trp_loc = h * (fx[:-1] + 2 * fm + fx[1:]) / 4  # trapezoidal rule
    simp_loc = h * (fx[:-1] + 4 * fm + fx[1:]) / 6  # Simpson rule
    est_loc = np.abs(simp_loc - trp_loc)  # Local error estimation
    err_tot = np.sum(est_loc)  # Estimate for quadrature error
    s = np.sum(simp_loc)  # Simpson approximation of integral value
    # Termination check
    if err_tot > rtol * np.abs(s) and err_tot > atol:
        refcells = np.flatnonzero(est_loc > 0.9 * np.sum(est_loc) / len(h))
        return adaptquad(f, np.sort(np.hstack([m, mp[refcells]])), rtol, atol)
    return s


def main():
    def f(x):
        return np.exp(-x**2)

    m = np.array([-100, 0.1, 0.5, 100])
    print('Sqrt(pi) - Integral of exp(-x^2) in domain [-100, 100] = {}'.format(
          np.sqrt(np.pi) - adaptquad(f, m, 1e-10, 1e-12)))


if __name__ == '__main__':
    main()
